blockchain/system/multicall.go

```diff
@@ -0,0 +1,80 @@
+	"github.com/klaytn/klaytn/accounts/abi/bind/backends"
+	"github.com/klaytn/klaytn/blockchain"
+	"github.com/klaytn/klaytn/blockchain/state"
+	"github.com/klaytn/klaytn/blockchain/types"
+	"github.com/klaytn/klaytn/blockchain/vm"
+	"github.com/klaytn/klaytn/common"
+	"github.com/klaytn/klaytn/contracts/contracts/system_contracts/multicall"
+)
+
+// ContractCallerForMultiCall is an implementation of ContractCaller only for MultiCall contract.
+// The caller interacts with a multicall contract on a read only basis.
+type ContractCallerForMultiCall struct {
+	state  *state.StateDB               // the state that is under process
+	chain  backends.BlockChainForCaller // chain containing the blockchain information
+	header *types.Header                // the header of a new block that is under process
+}
+
+func (caller *ContractCallerForMultiCall) CodeAt(ctx context.Context, contract common.Address, blockNumber *big.Int) ([]byte, error) {
+	return MultiCallCode, nil
+}
+
+// CallContract injects a multicall contract code into the state and executes the call.
+func (caller *ContractCallerForMultiCall) CallContract(ctx context.Context, call klaytn.CallMsg, blockNumber *big.Int) ([]byte, error) {
+	gasPrice := big.NewInt(0) // execute call regardless of the balance of the sender
+	gasLimit := uint64(1e8)   // enough gas limit to execute multicall contract functions
+	intrinsicGas := uint64(0) // read operation doesn't require intrinsicGas
+
+	// call.From: zero address will be assigned if nothing is specified
+	// call.To: the target contract address will be assigned by `BoundContract`
+	// call.Value: nil value is acceptable for `types.NewMessage`
+	// call.Data: a proper value will be assigned by `BoundContract`
+	// No need to handle access list here
+
+	// Set the code of the multicall contract
+	err := caller.state.SetCode(MultiCallAddr, MultiCallCode)
+	if err != nil {
+		return nil, err
+	}
+
+	msg := types.NewMessage(call.From, call.To, caller.state.GetNonce(call.From),
+		call.Value, gasLimit, gasPrice, call.Data, false, intrinsicGas, nil)
+
+	blockContext := blockchain.NewEVMBlockContext(caller.header, caller.chain, nil)
+	txContext := blockchain.NewEVMTxContext(msg, caller.header)
+	txContext.GasPrice = gasPrice                                                                // set gasPrice again if baseFee is assigned
+	evm := vm.NewEVM(blockContext, txContext, caller.state, caller.chain.Config(), &vm.Config{}) // no additional vm config required
+
+	result, err := blockchain.ApplyMessage(evm, msg)
+	return result.Return(), err
+}
+
+// NewMultiCallContractCaller creates a new instance of ContractCaller for MultiCall contract.
+// The state should be a new instance of StateDB obtained from StateAt() or State().
+func NewMultiCallContractCaller(state *state.StateDB, chain backends.BlockChainForCaller, header *types.Header) (*multicall.MultiCallContractCaller, error) {
+	c := &ContractCallerForMultiCall{state, chain, header}
```

Comment:
Does this not work?

```suggestion
func NewMultiCallContractCaller(chain backends.BlockChainForCaller, header *types.Header) (*multicall.MultiCallContractCaller, error) {
	state := chain.StateAt(header.Root)
	c := &ContractCallerForMultiCall{state, chain, header}
```

Reply:
It works and I think your suggestion is more safer. Thanks. Applied at [fc37a84](https://github.com/klaytn/klaytn/pull/2158/commits/fc37a846c6ebfa524e33a825155de8a5b25f1f99)